docs/source/reference/deprecation.rst

```diff
@@ -495,3 +495,46 @@ Schedule
 - In Numba 0.57: Support for CUDA toolkit 10.2 was removed.
 - In Numba 0.58: Support for CC < 5.0 and CUDA toolkits 11.0 and 11.1 will be
   removed.
+
+Deprecation of old-style ``NUMBA_CAPTURED_ERRORS``
+==================================================
+
+The old style of the ``NUMBA_CAPTURED_ERRORS`` environment variable is being 
+deprecated in Numba.
+
+Previously, this variable allowed controlling how Numba handles exceptions 
+during compilation that do not inherit from ``numba.core.errors.NumbaError``. 
+The default "old_style" behavior was to capture and wrap these errors, often 
+obscuring the original exception.
+
+The new "new_style" option treats non-``NumbaError`` exceptions as hard errors, 
+propagating them without capturing. This differentiates compilation errors from 
+unintended exceptions during compilation.
+
+The old style will eventually be removed in favor of the new behavior. Users 
+should migrate to setting ``NUMBA_CAPTURED_ERRORS='new_style'`` to opt-in to the 
+new exception handling. This will become the default in the future.
+
+Recommendations
+---------------
+
+- Projects that depends on Numba should set 
+  ``NUMBA_CAPTURED_ERRORS='new_style'`` for testing to find all places where 
+  non-``NumbaError`` exceptions are raised during compilation.
+- Replace any code that raises non-``NumbaError`` to indicate compilation error.
+  For instance, instead of raising ``TypeError``, raise 
+  ``numba.core.errors.NumbaTypeError``.
+
+
+Schedule
+--------
+
+- In Numba 0.58: support for ``NUMBA_CAPTURED_ERRORS='old_style'`` is 
+  deprecated.
+- In Numba 0.59: the use of ``NUMBA_CAPTURED_ERRORS='old_style'`` will raise 
+  deprecation warnings.
+- In Numba 0.60: support for ``NUMBA_CAPTURED_ERRORS='new_style'`` becomes the
+  default.
+- In Numba 0.61: support for ``NUMBA_CAPTURED_ERRORS='old_style'`` will be 
+  removed.
```

Comment:
Given that the documentation in 0.58 might essentially be missed in a lot of cases and need the warning being raised before it starts being noticed, perhaps we can shorten this cycle by one release if we start raising warnings in 0.58, giving:

```suggestion
- In Numba 0.58: support for ``NUMBA_CAPTURED_ERRORS='old_style'`` is 
  deprecated and will raise deprecation warnings.
- In Numba 0.59: support for ``NUMBA_CAPTURED_ERRORS='new_style'`` becomes the
  default.
- In Numba 0.60: support for ``NUMBA_CAPTURED_ERRORS='old_style'`` will be 
  removed.
```

Reply:
When discussing this with @stuartarchibald, we talked about making 0.58 raise warnings when old_style capturing is used but not when the env-var is set to `old_style` which is the default.  Maybe in 0.59, we can differentiate between setting old_style explicitly vs leaving it unset and only add warnings if the env-var is set to `old_style`.

Also, we usually give downstream 6months notice before changing behaviors.

Reply:
@gmarkall, i've pushed changes for this at https://github.com/numba/numba/pull/9090/commits/60b8027f90e63e2b9b591eb24627687b0df7c517

Reply:
Understood - thanks for the clarifications.